### DISCUSSION TITLE
Increase the usage of augmented assignment statements

### DISCUSSION THREAD

**Discussion by elfring** [networkx/networkx#5162] [Ideas]
:eyes: Some source code analysis tools can help to find opportunities for improving software components.
:thought_balloon: I propose to [increase the usage of augmented assignment statements](https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements "Augmented assignment statements") accordingly.

Would you like to integrate anything from a transformation result which can be generated by a command like the following?
(:point_right: Please check also for questionable change suggestions because of an evolving search pattern.)
```
[Markus_Elfring@fedora lokal]$ perl -p -i.orig -0777 -e 's/^(?<indentation>\s+)(?<target>\S+)\s*=\s*\k<target>\s*(?<operator>[+\-%&|^@]|\*\*?|\/\/?|<<|>>)/$+{indentation}$+{target} $+{operator}=/gm' $(find ~/Projekte/networkx/lokal -name '*.py')
```

**Comment by rossbar** [comment]
Hey @elfring , I converted this from an issue to a discussion as I thought it might be a slightly better fit here - I hope you don't mind!

At face value I don't think there'd be much resistance to using augmented assignment statements more widely as long as it improves the code, particularly with respect to readability (most important) or performance.

There are things to be careful of that jump out off the top of my head. For example, numpy treats many of the augmented assignment operators to mean "do this operation in-place". If the operand is of a different type than the dtype of the augmented array, this can result in errors:

```
>>> a = np.array([1, 2, 3], dtype=int)
>>> a += 5.0
Traceback (most recent call last)
   ...
UFuncTypeError: Cannot cast ufunc 'add' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```

NetworkX uses numpy arrays internally in many places, so I'm sure there are instances where this will cause problems. Running the modification script in the OP and then running the test suite on the modified code gives 45 errors, many of which I suspect fall into this category.

That said, I'm sure there are many other instances where such a change is an improvement. I don't have a strong opinion either way. I'm always +1 for improving code, though the degree of improvement for changes like this is arguably relatively minimal. The churn doesn't appear to be that bad, though the PR would require careful review. If there were a more robust automated conversion tool (i.e. one that took into account operator overloading) that might make this easier to review.

**Reply by elfring** [reply]
Thanks for such a constructive feedback.

:thought_balloon:
* The shown transformation approach can be improved in some ways, can't it?
* Would you like to mark any source code places where in-place modification should not be performed so far?

**Comment by rossbar** [comment]
> The shown transformation approach can be improved in some ways, can't it?

I'm sure it can, though I haven't thought hard about how :). One obvious thing would be to have a script that finds instances like the above (where types are incompatible) and skips or flags them. This would require some form of code instrospection though, which may significantly slow things down. 

> Would you like to mark any source code places where in-place modification should not be performed so far?

I don't think this is necessary - the test suite should (hopefully) catch such instances.

**Comment by tupui** [comment]
Have a read at this https://github.com/PyCQA/pylint/issues/5174 and other repos.

**Reply by rossbar** [reply]
Thanks for the heads up @tupui . The robots got me "/

**Reply by stefanv** [reply]
:robot:
